utils/sites.js

```diff
@@ -24,60 +23,121 @@ const submitParticipantsData = async (req, res, site) => {
         }
+
         siteCode = authorized.siteCode;
     }
 
-    if(req.body.data === undefined) return res.status(400).json(getResponseJSON('data is undefined in request body.', 400));
-    if(req.body.data.length === undefined || req.body.data.length < 1) return res.status(400).json(getResponseJSON('data array doesn\'t have any element.', 400));
+    if(req.body.data === undefined) return res.status(400).json(getResponseJSON('Bad request. Data is not defined in request body.', 400));
+    if(!Array.isArray(req.body.data)) return res.status(400).json(getResponseJSON('Bad request. Data must be an array.', 400));
+    if(req.body.data.length === undefined || req.body.data.length < 1) return res.status(400).json(getResponseJSON('Bad request. Data array does not have any elements.', 400));
+    if(req.body.data.length > 499) return res.status(400).json(getResponseJSON('Bad request. Data contains more than acceptable limit of 500 records.', 400));
 
-    if(req.body.data.length > 499) return res.status(400).json(getResponseJSON('More than acceptable limit of 500 records.', 400));
+    console.log(req.body.data);
+    
+    const dataArray = req.body.data;
 
-    const data = req.body.data;
-    console.log(`${JSON.stringify(data)}`);
+    let responseArray = [];
     let error = false;
-    let errorMsgs = [];
-    for(let obj of data){
-        if(obj.token){
-            const participantToken = obj.token;
-            const { getParticipantData } = require('./firestore');
-            const record = await getParticipantData(participantToken, siteCode);
-            if(record){
-                const docID = record.id;
-                delete obj.token;
-                let newStateElements = {}
-                for(let key in obj) {
-                    if(record.data.state[key] === undefined){
-                        newStateElements[`state.${key}`] = obj[key];
+
+    //let errorMsgs = [];
+
+    for(let dataObj of dataArray){
+        if(dataObj.token === undefined) {
+            error = true;
+            responseArray.push({'Invalid Request': {'Token': 'UNDEFINED', 'Errors': 'Token not defined in data object.'}});
+            continue;
+        }
+        
+        const participantToken = dataObj.token;
+        const { getParticipantData } = require('./firestore');
+        const record = await getParticipantData(participantToken, siteCode);
+
+        if(!record) {
+            error = true;
+            responseArray.push({'Invalid Request': {'Token': participantToken, 'Errors': 'Token does not exist.'}});
+            continue;
+        }
+
+        const docID = record.id;
+        const docData = record.data;
+
+        const flat = (obj, att, attribute) => {
+            for(let k in obj) {
+                if(typeof(obj[k]) === 'object') flat(obj[k], att, attribute ? `${attribute}.${k}`: k)
+                else flattened[att][attribute ? `${attribute}.${k}`: k] = obj[k]
+            }
+        }
```

Comment:
The function can be moved out of the loop, so that it doesn't need to be re-defined in each iteration.

Reply:
Noted